src/analyses/base.ml

```diff
@@ -1466,16 +1466,33 @@ struct
         (* | `Address a, `Address b -> ... *)
         | a1, a2 -> fallback ("binop: got abstract values that are not `Int: " ^ sprint VD.pretty a1 ^ " and " ^ sprint VD.pretty a2))
       | Lval x -> (* meet x with c *)
-        let c' = match typeOfLval x with
+        let t = Cil.unrollType (typeOfLval x) in  (* unroll type to deal with TNamed *)
+        let c' = match t with
           | TPtr _ -> `Address (AD.of_int (module ID) c)
-          | TInt (ik, _) -> `Int (ID.cast_to ik c )
+          | TInt (ik, _)
+          | TEnum ({ekind = ik; _}, _) -> `Int (ID.cast_to ik c )
           | _ -> `Int c
         in
         let oldv = eval (Lval x) in
-        let v = VD.meet oldv c' in
-        if is_some_bot v then raise Deadcode
-        else
-          if M.tracing then M.tracel "inv" "improve lval %a = %a with %a (from %a), meet = %a\n" d_lval x VD.pretty oldv VD.pretty c' ID.pretty c VD.pretty v;
+        let skip = if Cil.isVoidPtrType t then
+            match oldv with
+            | `Address _ -> false
+            | _ -> true (* We do not change type on a cast to void*, so we can not meet here without type errors*)
```

Comment:
Yes, but only because
```
custom_t ci;
```

If we have more information for ci, we will also have it inside of `if(!ptr) {...`.

`VD.meet` can meet `Int` and `Address`

https://github.com/goblint/analyzer/blob/master/src/cdomains/valueDomain.ml#L484 
The caveat is that the ranges will not match here and we will warn.

The problem with creating `c'` of the appropriate type for the abstract value is that we don't have the `ikind` readily available.

Reply:
Also, one might wonder if this is the right thing to do anyway, as the C standard leaves it up to the implementation what happens on a cast of an integer type to a pointer type.

C99: 6.3.2.3
```
5 An integer may be converted to any pointer type. Except as previously specified, the result is implementation-defined, might not be correctly aligned, might not point to an entity of the referenced type, and might be a trap representation.56)
```

Reply:
> If we have more information for ci, we will also have it inside of if(!ptr) {...

Well, that's always true, i.e. 'invariant' will only improve values.

> one might wonder if this is the right thing to do anyway, as the C standard leaves it up to the implementation

What else should you do? What example are you referring to? We only want to make sure that we get a `null_ptr`/`not_null` - there's not much more one can do, but this should work.
Casts to `(void*)` are either used to do some byte-level stuff or to get around the type system and do some parametric polymorphism at runtime. In this case the value is always cast back and we don't want to lose the original abstract value.

Reply:
> The problem with creating c' of the appropriate type for the abstract value is that we don't have the ikind readily available.

If we have `Cil.isVoidPtrType t` and `oldv = Address` everything works already.
For `oldv = Int _` we now don't do anything, but we could just create an `Int c`.
The `ikind` could be top since the meet would get rid of it, but if you want to avoid your same ikind warning, you'd have to get it from the `oldv`.

Reply:
> Casts to (void*) are either used to do some byte-level stuff or to get around the type system and do some parametric polymorphism at runtime. In this case the value is always cast back and we don't want to lose the original abstract value.

It's just that casts to `(void*)` from an integer types and then back to an integer type are not guaranteed to yield the original value in C unless you are talking about the constant `0`.
Now, for other pointer types there is of course no issue.

Reply:
> The ikind could be top since the meet would get rid of it
> but if you want to avoid your same ikind warning, you'd have to get it from the oldv.
Yes, this would necessitate some sort of method that gets an ikind from an `Int` (or creates an int of the same `ikind`) which is also sort of ugly. I was hoping for a better solution there. :)

Reply:
> It's just that casts to (void*) from an integer type t and then back to t are not guaranteed to yield the original value in C unless you are talking about the constant 0.

Should investigate this. I guess this may only happen for integer types bigger than the pointer type.
Potentially problematic on <64 bit platforms: https://de.wikipedia.org/wiki/Datentypen_in_C#Datenmodell

Reply:
> I was hoping for a better solution there. :)

Just meet and remove the check there?

Reply:
> Just meet and remove the check there?

The normal meet does not cut it here, it will lead to the ranges being wrong later.
The case where it breaks is for `unsigned int` (which is also the one in the example).
`meet(Excluded({},[0,64]), Excluded({0},[-63,63]) = Excluded({0},[0,63]) `

Reply:
I think I found a quite nice solution now:

- On cast to `void*` for integers, cast them to `TInt(ptr_ikind (),[])` where `ptr_ikind()` comes from CIL.
- When meeting `Int` and `Address`, cast the address to `TInt(ptr_ikind (),[])`
- This allows removing the special handling of `void*` in `invariant`

Reply:
> meet(Excluded({},[0,64]), Excluded({0},[-63,63]) = Excluded({0},[0,63])

That means the `top_range` is not big enough. Might be a problematic elsewhere too.
If you change it to be big enough to encompass all int types, e.g. `[-129,129]`, it would work.
This would encode a top for range, which is something else than some known range or an over-approximation of ranges. Meeting with top should always be ok and just give you the other value. Also, it would only be used here to get around the fact that there's currently no function to get the ikind out of an abstract int.

Reply:
> - When meeting Int and Address, cast the address to TInt(ptr_ikind (),[])

This loses precision for addresses, no? I think we always want to keep addresses since we get unsound if we throw them away (might be some function that's called).

> - On cast to void* for integers, cast them to TInt(ptr_ikind (),[]) where ptr_ikind() comes from CIL.

Does this lose precision? Negative values will overflow on cast since `ptr_ikind` is `unsigned long`.

Reply:
>This loses precision for addresses, no? I think we always want to keep addresses since we get unsound if we throw them away (might be some function that's called). 

This is what we do currently in this case:
https://github.com/goblint/analyzer/blob/a7e04ee071c56980ab651c9baff91d55acf42a2d/src/cdomains/valueDomain.ml#L484

vs the proposed code:
https://github.com/goblint/analyzer/blob/168e5ade31cd44938362ed3cedcf19c7c5ee1525/src/cdomains/valueDomain.ml#L499

All that changes is the `ikind` we cast to.

Reply:
> Does this lose precision? Negative values will overflow on cast since ptr_ikind is unsigned long.
Yes, this is true. But could be fixed for some domains (e.g. the intervals come to mind) by having a better `cast`.
What I like about this solution is that it seems clean to me.

Reply:
Ok, the second is better, I'm just confused that we don't keep the address.
Idk if this is the right place (ideally such meets of different abstract values should go away), but it should probably check if there are lvals inside the address set and if so cast the int to an address instead (which will add/remove null). Or more generally: always change to the type of the more precise value.

Reply:
> by having a better `cast`

I don't remember, might be that it can already deal with casting to unsigned and back without loss.

Reply:
> ideally such meets of different abstract values should go away

It is unfortunately precisely such hacks such as preserving an `Int` through a cast to `void*` that cause the `typ` that CIL knows to not match our abstract types, which in turn means CIL doesn't insert casts, which is why we end up doing things such as meeting `Int` and `Address`.
Another thing that is strange about the meet here is that it is not commutative:
https://github.com/goblint/analyzer/blob/a7e04ee071c56980ab651c9baff91d55acf42a2d/src/cdomains/valueDomain.ml#L484-L485

But I guess these thing should be discussed seperatly.

Reply:
I'm open to better solutions. Helmut once proposed these 'flavored' ints which would merge those things.
I guess the reason for it being non-commutative is that it should keep the old type (left) and just try to improve it.
Maybe better to do this conversion in `invariant` and make it commutative in ValueDomain and fail if it's different types there.